### ISSUE TITLE
the optimizer step problem

### ISSUE THREAD

**Issue by JoJoistheBestOne** [JaydenLyh/Reward-Forcing#7]
hi authors, recently ,i noticed in the TRAIN_GENERATOR, it seems this will update twice?

condition1 is `TRAIN_GENERATOR = self.step % (self.config.dfake_gen_update_ratio * self.gradient_accumulation_steps) == 0`

condition2 is `if self.step % self.gradient_accumulation_steps == 0:`

when condition1 is qulified, and the condition2 is qulified also.
then this updating of `self.generator_optimizer.step()` run twice?

https://github.com/JaydenLyh/Reward-Forcing/blob/main/trainer/rewarded_distillation.py#L360-L363

**Comment by JoJoistheBestOne** [comment]
in addition, it seems like the grad is accumulating and applied to the new-updated-model for each step.

https://github.com/JaydenLyh/Reward-Forcing/blob/main/trainer/rewarded_distillation.py#L348-L354

**Comment by JoJoistheBestOne** [comment]
to summarize,

the `gradient_accumulation_steps` and the `dfake_gen_update_ratio` conflicts, and the `critic_optimizer` will update at the old-gradient because there is not zero_grad().

it will:
1) the `Generator` will ONLY update at  gradient_accumulation_steps*dfake_gen_update_ratio -1
2) the `Critict` will update on the old grad.

then I personally recommend this code segment.


```
            # 1. Logic for gradient accumulation and update frequency
            update_idx = self.step // self.gradient_accumulation_steps
            TRAIN_GENERATOR = (update_idx % self.config.dfake_gen_update_ratio == 0)
            
            is_accum_start = (self.step % self.gradient_accumulation_steps == 0)
            is_accum_end = (self.step % self.gradient_accumulation_steps == self.gradient_accumulation_steps - 1)

            # 2. Start of accumulation cycle: Zero gradients
            if is_accum_start:
                if TRAIN_GENERATOR:
                    self.generator_optimizer.zero_grad(set_to_none=True)
                self.critic_optimizer.zero_grad(set_to_none=True)

            # 3. Train the generator (if in generator update cycle)
            if TRAIN_GENERATOR:
                batch = next(self.dataloader)
                generator_log_dict = self.fwdbwd_one_step(batch, True)

            # 4. Train the critic (always participates in accumulation)
            batch = next(self.dataloader)
            critic_log_dict = self.fwdbwd_one_step(batch, False)

            # 5. End of accumulation cycle: Perform optimizer step
            if is_accum_end:
                if TRAIN_GENERATOR:
                    self.generator_optimizer.step()
                    if self.generator_ema is not None:
                        self.generator_ema.update(self.model.generator)
                
                self.critic_optimizer.step()

            # physical step increment
            self.step += 1

```

**Comment by Zian-0427** [comment]
I agree. So is the current public checkpoint trained with this setting? Seems odd.
